### DISCUSSION TITLE
Support streaming in DuckDB

### DISCUSSION THREAD

**Discussion by code1704** [duckdb/duckdb#5223] [General]
Is there any plan to support live streaming data in duckdb?

**Comment by Mytherin** [comment]
Could you clarify what you mean/envision with live streaming data?

**Reply by code1704** [reply]
In quantitative trading, there's new stock price data coming every few ticks.

1. to receive the live data, e.g. open/high/low/close/volume every second
2. calculate the factors when data comes in, e.g. SELECT returns=close/open FROM xxx, then append the data to some view or push to some client

(I think we can do that in flink or influxdb?)

**Reply by Alex-Monahan** [reply]
DuckDB's engine is designed to process batches of data at a time really efficiently rather than 1 row at a time. The vector size is now 2048 items, so we prefer to process data that is at least that large all at once. Given that, do you still think it would be a good fit for what you are aiming for? What system do you use to do this work today? Could you provide a few more details around how many rows each query is operating on?

**Reply by code1704** [reply]
Thanks for your reply.

1. YES, duckdb works good for batch processing. we use duckdb on backtesting (offline) (History data: about 10TB+, saved as arrow files)
2. And we want to use it in realtime trading (online) as well, so that we do not need to rewrite some of the code. We do not have a good solution yet (when we receive the trading data every second, how to send to duckdb? create a table in memory? trigger the duckdb processing?). e.g. data api: https://github.com/binance/binance-spot-api-docs/blob/master/web-socket-streams.md#trade-streams

```
streaming_data = ??
for row in duckdb.query("SELECT high/low FROM streaming_data"):
    ...
```
or create materialized view https://materialize.com/docs/sql/create-materialized-view/ ?

**Comment by fabianoliver** [comment]
Just to piggyback off this existing discussion: Having the ability to continuously stream query results as the underlying data changes could be enormously useful!

I think there's a gaping niche in terms of software that allows to aggregate different datastreams in a SQL table-like manner, while being purely in-process, lightweight and fast.

There are a few frameworks that serve a similar purpose, but they're all geared towards much, much more big-data oriented workloads. Flink's table API is a prime example. Absolutely amazing tool, but it has tremendous overhead, and is realistically geared towards continuously running queries spread across multiple nodes with heaps of CPU and memory.

I very frequently need to do similar operations, but operating purely in-progress for relatively short-lived ad hoc user requests: "Some application subscribes to a few data streams, must aggregate these in-process and in-memory, be able to set up the pipeline & calculate results with a few tens to a few hundreds of milliseconds, and stream back result sets as long as the client is subscribed".
There are plenty of streaming APIs from Rx to others, and you can do all of that with them, but none of them is out of the box geared towards table-like data / table-like operations (it's possible to implement that kind of functionality on top of them, but it's not entirely trivial).

Now say for example, you were able to create a bunch of in-memory tables in DuckDb, then perform some form of streaming query that derives some results from these tables. Every time the source tables are modified (rows deleted, added, changed), that should eventually feed through to your ongoing streaming query, and publish some delete/add/change message to whatever row of your derived resultset has been affected.

I'm unfortunately not familiar with DuckDb's internals, and in how far there's (internal) observability of changes. Is that something that would generally be feasible?
